keps/sig-node/20200303-shared-pod-limits.md

```diff
@@ -0,0 +1,411 @@
+        = ... =
+          |
+          +-- containerN (N-th container)
+</pre>
+
+The current implementation will set the cgroup limits for the memory/CPU resources on the pod level of the hierarchy only in the following cases:
+1. The QoS level is `Guaranteed`. 
+1. The QoS level is `Burstable` and all of the containers specify a limit for the relevant resource. 
+
+For the current definition of the `Guaranteed` QoS level this proposal would not make any modifications. Since each container is provided with its
+own limit, defining a pod-level limit is redundant. If the pod level limit request is smaller than the sum of limit for all the containers, then 
+the limit requested by the containers is unreachable. Conversely if the pod level limit request is larger than the sum of the limit for all the 
+containers then the excess is not relevant since the individual containers can never use more than their requested share, and even if all of the 
+containers use the maximum allowed by their defined limit, it can never amount to the pod-level limit.
+
+For pods in the `Burstable` and `BestEffort` QoS levels specifying a resource limit on the pod level means that resources can be shared across containers.
+Sharing resources across containers is possible only if individual containers release the resources that are not being used. For CPU this is trivial.
+For memory, this depends on the application being run actually releasing the memory back to the operating system. See [below](#Memory-Reuse-Across-cgroups)
+for an analysis.
+
+The proposal in this KEP is to allow users to define resources on the Pod level itself. Only limits will be allowed on this level so that the effect on
+the scheduler algorithm will be minimal. The scheduler will continue to use the resource requests defined on the container level as its input. Kubelet 
+will configure the pod-level cgroups as defined on the pod level. If the pod level doesn't include any resource limits, then kubelet will continue to
+function as it does today.
+
+For example, consider the following Pod:
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  labels:
+    run: nginx
+  name: nginx
+spec:
+  resources:
+    limits:
+      memory: 384M
+      cpu: "2"
+  containers:
+  - image: proxy
+    name: envoy:latest
+  - name: nginx
+    image: nginx:latest
+    command: [ "/usr/bin/tail", "-f", "/dev/null"] 
+    resources:
+      requests:
+        memory: 128M
+        cpu: "0.5"
+      limits:
+        memory: 256M
+        cpu: "1"
+```
+
+The cgroup hierarchy for each of these resources (memory and cpu) would be this:
+
+<pre>
+  QoS CGroup (one of guaranteed, burstable, or besteffort)
+   |
+   \ pod (memory: 384M limit, CPU: 2 core)
+      |
+      +-- container0 (pause container, memory: unlimited, cpu: unlimited quota)
+      |
+      +-- container1 (proxy container, memory: unlimited, CPU: unlimited quota)
+      |
+      +-- container2 (nginx container, memory: 256M limit, CPU: 1 core)
+</pre>
```

Comment:
I keep thinking what is the best mental model, and how people would understand this. I think most people thinks that if you set a `request`, you are guaranteed to get _at least_ that much of a resource, and therefore the tree will look something like this:

```

  QoS CGroup (one of guaranteed, burstable, or besteffort)
   |
   \ pod (memory: 384M limit, CPU: 2 core)
      |
      +-- container0 (pause container, memory: 128M shared with container1, cpu: 1 core shared with container1)
      |
      +-- container1 (proxy container, memory: 128M shared with container0, CPU: 1 core shared with container0)
      |
      +-- container2 (nginx container, memory: 256M limit, CPU: 1 core)
```
This would however require some kind of qos hierarchy for the containers in the pod, and that is probably not realistic, and would overcomplicate stuff quite a bit.

This should however be properly documented in the docs to make sure that users understand how it works. (and yeah, you essentially not guaranteed anything with todays model either, but the OOM etc. can help you with that, but it will not help in this specific case, inside a pod).

This is another reason why I think pod-level requests should be a thing as well, since you probably want that example-pod to be burstable, (or even guaranteed in some cases).

Reply:
The only effects that setting the `Request` field has are as a parameter for the scheduler, and for choosing a QoS level. The Linux kernel itself doesn't support resource reservation as part of the memory or CPU controller groups mechanism. A pod can specify a `Request` field equal to a `Limit` field, be assigned to the `Guaranteed` QoS level, and still fail to allocate the amount of memory that it requested. 

I completely agree that the documentation should make this point clear. 

Memory reservation is a side-effect that actually works if and only if **all** of the pods on a node are `Guaranteed` and no over-provisioning is allowed. Otherwise, being in the `Guaranteed` QoS level in the above example will only cause the kubelet (if under memory pressure) to try to evict some other pod in another QoS level first. 

I agree that it makes sense to allow these semantics to hold true also for pod-level requests. So if a pod-level `Request` equals the pod level `Limit`, the pod should be in the `Guaranteed` QoS level. Also, if a pod has pod-level resources, that would automatically place it in the `Burstable` QoS level, regardless of any limits set on the container level.

I will update the KEP.